### PR TITLE
Crosswalk Issue Ignore / Force Resolution

### PR DESCRIPTION
**Pull request by devin-mccurdy**
## Description

Crosswalk issues are considered resolved if the Weams, IpedsHD, and Crosswalk records all match, but there will be cases where users will need to manually mark a crosswalk issue as resolved.  These ignored issues need to removed from following CrossWalk issue rebuilds.

DB migration PR: https://github.com/department-of-veterans-affairs/gibct-data-service/pull/593

Issue: https://app.zenhub.com/workspaces/vft-59c95ae5fda7577a9b3184f8/issues/department-of-veterans-affairs/va.gov-team/5782

https://app.zenhub.com/workspaces/vft-59c95ae5fda7577a9b3184f8/issues/department-of-veterans-affairs/va.gov-team/6257

## Testing done
Tested locally and QA reviewed

## Screenshots
image

## Acceptance criteria
- [x] The user has the ability to mark a crosswalk issue entry as resolved, so that it no longer shows up in the crosswalk issues table.
- [x] If the user uploads a new file (WEAMS, Crosswalk, IPEDSHD) and neither the IPED nor OPE ID have changed, a crosswalk issue will not be added to the table if it was previously marked as resolved.
- [x] If the user uploads a new file (WEAMS, Crosswalk, IPEDSHD) and either the IPED or OPE ID have changed, a crosswalk issue will be added to the table.

## Definition of done
- [x] Events are logged appropriately
- [x] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

**Review by shawkey-daniel — APPROVED** [review]
LGTM

**Review by sonntag-adam — APPROVED** [review]

**Review by philipwritescode — DISMISSED** [review]
This looks good and I would approve but I don't want the html to get mixed up if that's a mistake

File: app/views/crosswalk_issues/show_partial.html.erb [hunk 1]
```diff
@@ -65,8 +65,8 @@
                               "#{Settings.college_navigator_url}/?id=#{@issue.ipeds_hd&.cross}",
                               target: :_blank
                   %>
-                <td/>
                 <td><%= @issue.ipeds_hd&.ope %></td>
+                <td/>
```
[the remark this answers is not in the document]
Comment (dmccurdybah) [reply app/views/crosswalk_issues/show_partial.html.erb]: Thanks!

[the remark this answers is not in the document]
Comment (jholton) [reply app/views/crosswalk_issues/show_partial.html.erb]: No comment as to whether it's supposed to be there, but it's not a nested `td` element.  It's `<td />`, not `</td>`.  i.e. it's a self closing tag, or an empty cell.

[the remark this answers is not in the document]
Comment (dmccurdybah) [reply app/views/crosswalk_issues/show_partial.html.erb]: Yes, it's supposed to be there, but there was also a missing </td> above which I believe was causing the confusion

**Review by jholton — COMMENTED** [review]
From the associated issue:
> If the user uploads a new file (WEAMS, Crosswalk, IPEDSHD) and either the IPED or OPE ID have changed, a crosswalk issue will be added to the table.

Can you add spec(s) demonstrating this scenario?

**Comment by jholton** [comment]
re:
> If the user uploads a new file (WEAMS, Crosswalk, IPEDSHD) and either the IPED or OPE ID have changed, a crosswalk issue will be added to the table.

```ruby
      def ignore_issue(issue) # copied from controller
        IgnoredCrosswalkIssue.create(
          cross: issue.ipeds_hd.present? ? issue.ipeds_hd.cross : issue.crosswalk.cross,
          ope: issue.ipeds_hd.present? ? issue.ipeds_hd.ope : issue.crosswalk.ope,
          facility_code: issue.weam.facility_code
        )
      end

      # Weam( facility_code: '99Z99999', cross: '888888', ope: '00000001' )
      create :weam, :ncd, :approved_poo_and_law_code, :crosswalk_issue_matchable_by_cross,
             :crosswalk_issue_matchable_by_facility_code
      # IpedsHd( cross: '888888', ope: nil )
      create :ipeds_hd, :crosswalk_issue_matchable_by_cross

      expect { described_class.rebuild }
        .to change { described_class.by_issue_type(CrosswalkIssue::PARTIAL_MATCH_TYPE).count }.from(0).to(1)

      # ignore the issue and delete it:
      issue = described_class.by_issue_type(CrosswalkIssue::PARTIAL_MATCH_TYPE).first
      ignore_issue(issue)
      issue.delete

      # issue is gone:
      expect(described_class.by_issue_type(CrosswalkIssue::PARTIAL_MATCH_TYPE).count).to eq 0

      # rebuild won't create an issue because it is ignored
      expect { described_class.rebuild }
        .not_to change { described_class.by_issue_type(CrosswalkIssue::PARTIAL_MATCH_TYPE).count }

      # Now let's simulate uploading a new IpedsHd file:
      IpedsHd.destroy_all

      # Record now has an `ope` set.  Since ope has changed, we expect an issue to be generated
      create :ipeds_hd, :crosswalk_issue_matchable_by_cross, ope: '00000001'
      expect { described_class.rebuild }
        .to change { described_class.by_issue_type(CrosswalkIssue::PARTIAL_MATCH_TYPE).count }.from(0).to(1)
```
Result:
> expected `described_class.by_issue_type(CrosswalkIssue::PARTIAL_MATCH_TYPE).count` to have changed from 0 to 1, but did not change

**Comment by devin-mccurdy** [comment]
> From the associated issue:
> 
> > If the user uploads a new file (WEAMS, Crosswalk, IPEDSHD) and either the IPED or OPE ID have changed, a crosswalk issue will be added to the table.
> 
> Can you add spec(s) demonstrating this scenario?

I'll ask for the AC to get updated because this is definitely confusing.

A CrossWalk issue would only be created when there is a valid match between Weams and IpedsHD using the existing matching criteria, so it's not the case that updating the IPED or OPE ID would always cause a CrosswalkIssue to be created when the table is rebuilt.

**Comment by jholton** [comment]
> A CrossWalk issue would only be created when there is a valid match between Weams and IpedsHD using the existing matching criteria, so it's not the case that updating the IPED or OPE ID would always cause a CrosswalkIssue to be created when the table is rebuilt.

First, I just realized that in my example above, at the beginning of the example I accidentally wrote 
`create :ipeds_hd, :crosswalk_issue_matchable_by_cross, ope: '00000001'`
when I meant
`create :ipeds_hd, :crosswalk_issue_matchable_by_cross`
I have fixed it.

To your comment:
Right.  I _think_ I understand (but maybe I don't!).  My fixed example reflects this.

Both of the following specs pass.  The only difference is the specification on an `ope` value on the `create :ipeds_hd` line:
```ruby
      create :weam, :ncd, :approved_poo_and_law_code, :crosswalk_issue_matchable_by_cross,
             :crosswalk_issue_matchable_by_facility_code
      create :ipeds_hd, :crosswalk_issue_matchable_by_cross

      expect { described_class.rebuild }
        .to change { described_class.by_issue_type(CrosswalkIssue::PARTIAL_MATCH_TYPE).count }.from(0).to(1)
```

```ruby
      create :weam, :ncd, :approved_poo_and_law_code, :crosswalk_issue_matchable_by_cross,
             :crosswalk_issue_matchable_by_facility_code
      create :ipeds_hd, :crosswalk_issue_matchable_by_cross, ope: '00000001'

      expect { described_class.rebuild }
        .to change { described_class.by_issue_type(CrosswalkIssue::PARTIAL_MATCH_TYPE).count }.from(0).to(1)
```

So, whether or not this particular `ipeds_hd` record is created with an `ope`, it results in a `CrosswalkIssue` of `PARTIAL_MATCH_TYPE` being created.

So in my scenario, you have a `weam` record
```ruby
create :weam, :ncd, :approved_poo_and_law_code, :crosswalk_issue_matchable_by_cross,
             :crosswalk_issue_matchable_by_facility_code
```
and an `ipeds_hd` record
```ruby
create :ipeds_hd, :crosswalk_issue_matchable_by_cross
```
that result in a `CrosswalkIssue::PARTIAL_MATCH_TYPE` record being created.
```ruby
expect { described_class.rebuild }
        .to change { described_class.by_issue_type(CrosswalkIssue::PARTIAL_MATCH_TYPE).count }.from(0).to(1)
```
Now the end user ignores that issue:
```ruby
      ignore_issue(issue)
      issue.delete
```

Now the end user creates (uploads) a new version of that `ipeds_hd` record with an `ope` value:
```ruby
create :ipeds_hd, :crosswalk_issue_matchable_by_cross, ope: '00000001'
```
Given that:
* We know that this new `ipeds_hd` record in conjunction with our `weams` record _does_ meet existing matching criteria for creating a `crosswalk_issues` record
* The user has uploade a new `ipeds_hd` record wherein the `ope` value has changed

Shouldn't a `crosswalk_issues` record be generated?

**Review by sonntag-adam — APPROVED** [review]

**Comment by devin-mccurdy** [comment]
> * The user has uploade a new `ipeds_hd` record wherein the `ope` value has changed
> 
> Shouldn't a `crosswalk_issues` record be generated?

Yes, sorry, I misunderstood your point. You're correct, this wasn't handling null OPE values correctly (even though I wouldn't think that we'd ever have a null OPE value in the IPEDSHD data, it allows nulls). Also, that's a much nicer way to write these tests.  Thanks!

**Review by jholton — APPROVED** [review]

**Pull request merged**